torch/lib/THD/master_worker/master/generic/THDTensorMath.cpp

```diff
@@ -0,0 +1,314 @@
+  THArgCheck(k > 0 && k <= sliceSize, 2, "k not in range for dimension");
+
+  THLongStorage *topKSize = THDTensor_(newSizeOf)(t);
+  THLongStorage_set(topKSize, dim, k);
+  THDTensor_(resize)(rt_, topKSize, NULL);
+  THDLongTensor_resize(ri_, topKSize, NULL);
+  THLongStorage_free(topKSize);
+
+  masterCommandChannel->sendMessage(
+    packMessage(Functions::tensorTopk, rt_, ri_, t, k, dim, dir, sorted),
+    THDState::s_current_worker
+  );
+}
+
+void THDTensor_(tril)(THDTensor *r_, THDTensor *t, long k) {
+  THArgCheck(THDTensor_(nDimension)(t) == 2, 1, "expected a matrix");
+
+  THDTensor_(resizeAs)(r_, t);
+
+  masterCommandChannel->sendMessage(
+    packMessage(Functions::tensorTril, r_, t, k),
+    THDState::s_current_worker
+  );
+}
+
+void THDTensor_(triu)(THDTensor *r_, THDTensor *t, long k) {
+  THArgCheck(THDTensor_(nDimension)(t) == 2, 1, "expected a matrix");
+
+  THDTensor_(resizeAs)(r_, t);
+
+  masterCommandChannel->sendMessage(
+    packMessage(Functions::tensorTriu, r_, t, k),
+    THDState::s_current_worker
+  );
+}
+void THDTensor_(cat)(THDTensor *r_, THDTensor *ta, THDTensor *tb, int dimension) {
+  THDTensor* inputs[2];
+  inputs[0] = ta;
+  inputs[1] = tb;
+  THDTensor_(catArray)(r_, inputs, 2, dimension);
+}
+
+void THDTensor_(catArray)(THDTensor *result, THDTensor **inputs,
+                          int numInputs, int dimension) {
+  THLongStorage *size;
+  long offset;
+  int ndim = dimension + 1;
+  int ldimension = dimension;
+  bool allEmpty = true;
+  for (int i = 0; i < numInputs; i++)
+    ndim = THMax(ndim, inputs[i]->nDimension);
+
+  if (dimension == -2)
+    ldimension = ndim ? (ndim - 1) : 0;
+
+  THArgCheck(numInputs > 0, 3, "invalid number of inputs %d", numInputs);
+  THArgCheck(ldimension >= 0, 4, "invalid dimension %d", dimension + TH_INDEX_BASE);
+
+  size = THLongStorage_newWithSize(ndim);
+
+  for (int i = 0; i < ndim; i++) {
+    long dimSize = i < inputs[0]->nDimension ? inputs[0]->size[i] : 1;
+    if (i == ldimension) {
+      for (int j = 1; j < numInputs; j++) {
+        dimSize += i < inputs[j]->nDimension ? inputs[j]->size[i] : 1;
+      }
+    } else {
+      for (int j = 1; j < numInputs; j++) {
+        long sz = (i < inputs[j]->nDimension ? inputs[j]->size[i] : 1);
+        if (dimSize != sz && dimSize && sz) {
+          THLongStorage_free(size);
+          THError("inconsistent tensor sizes");
+        } else if (!dimSize) {
+          dimSize = sz;
+        }
+      }
+    }
+    allEmpty = allEmpty && !dimSize;
+    size->data[i] = dimSize;
+  }
+
+  if (!allEmpty) {
+    THDTensor_(resize)(result, size, NULL);
+    std::vector<THDTensor*> inputs_vec;
+    inputs_vec.assign(inputs, inputs + numInputs);
+
+    // There's no need to send numInputs,
+    // since sending inputs_vec does this implicitly
+    masterCommandChannel->sendMessage(
+      packMessage(Functions::tensorCatArray, result, inputs_vec, dimension),
+      THDState::s_current_worker
+    );
+  }
+
```

Comment:
It only narrows some temporary tensor to copy `inputs` into result in an appropriate way - so it doesn't touch the metadata